bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/internal/discovery/DiscoveryProtocol.java

```diff
@@ -0,0 +1,84 @@
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.broadlink.internal.discovery;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.binding.broadlink.internal.BroadlinkProtocol;
+import org.openhab.binding.broadlink.internal.NetworkUtils;
+import org.openhab.binding.broadlink.internal.socket.BroadlinkSocket;
+import org.openhab.binding.broadlink.internal.socket.BroadlinkSocketListener;
+import org.slf4j.Logger;
+
+/**
+ * @author John Marshall - Initial contribution
+ */
+@NonNullByDefault
+public class DiscoveryProtocol {
+
+    private static class AsyncDiscoveryThread extends Thread {
+        private final BroadlinkSocketListener listener;
+        private final long timeoutMillis;
+        private final DiscoveryFinishedListener finishedListener;
+        private final Logger logger;
+
+        AsyncDiscoveryThread(BroadlinkSocketListener listener, long timeoutMillis,
+                DiscoveryFinishedListener finishedListener, Logger logger) {
+            this.listener = listener;
+            this.timeoutMillis = timeoutMillis;
+            this.finishedListener = finishedListener;
+            this.logger = logger;
+        }
+
+        @Override
+        public void run() {
+            BroadlinkSocket.registerListener(listener, logger);
+            DiscoveryProtocol.discoverDevices(logger);
+            DiscoveryProtocol.waitUntilEnded(timeoutMillis, logger);
+            logger.warn("Ended Broadlink device scan...");
+            BroadlinkSocket.unregisterListener(listener, logger);
+            finishedListener.onDiscoveryFinished();
+        }
+    }
+
+    public static void beginAsync(BroadlinkSocketListener listener, long discoveryTimeoutMillis,
+            DiscoveryFinishedListener discoveryFinishedListener, Logger logger) {
+        logger.warn("Beginning async Broadlink device scan; will wait {} ms for responses", discoveryTimeoutMillis);
+        AsyncDiscoveryThread adt = new AsyncDiscoveryThread(listener, discoveryTimeoutMillis, discoveryFinishedListener,
+                logger);
+        adt.start();
+    }
+
+    public static void discoverDevices(Logger logger) {
+        try {
+            InetAddress localAddress = NetworkUtils.getLocalHostLANAddress();
+            int localPort = NetworkUtils.nextFreePort(localAddress, 1024, 3000);
+            byte message[] = BroadlinkProtocol.buildDiscoveryPacket(localAddress.getHostAddress(), localPort);
+            BroadlinkSocket.sendMessage(message, "255.255.255.255", 80, logger);
+        } catch (UnknownHostException e) {
+            logger.warn("Failed to initiate discovery", e);
+        }
+    }
+
+    private static void waitUntilEnded(long discoveryTimeoutMillis, Logger logger) {
+        try {
+            logger.warn("Broadlink device scan waiting for {} ms to complete ...", discoveryTimeoutMillis);
```

Comment:
Can it be removed? warn seems wrong here, debug or trace levels unless it actually causes an issue.
Please can you do a FIND on your files for `logger.warn` and review yourself if they need to be removed, or changed to another level like debug or trace if they create a lot of lines?

Guide for logging is here and covers when each level should be used.

https://www.openhab.org/docs/developer/guidelines.html#f-logging

Reply:
Cleaned up all the logger.warn statements.